utilities/backup/backup_engine.cc

```diff
@@ -2353,8 +2495,19 @@ IOStatus BackupEngineImpl::AddBackupFileWorkItem(
     BackupAfterCopyOrCreateWorkItem after_copy_or_create_work_item(
         copy_or_create_work_item.result.get_future(), shared, need_to_copy,
         backup_env_, temp_dest_path, final_dest_path, dst_relative);
-    files_to_copy_or_create_.write(std::move(copy_or_create_work_item));
-    backup_items_to_finish.push_back(std::move(after_copy_or_create_work_item));
+    if (excludable_items != nullptr && shared && shared_checksum &&
+        need_to_copy) {
+      ROCKS_LOG_INFO(options_.info_log, "Copying (if not excluded) %s to %s",
+                     fname.c_str(), copy_dest_path->c_str());
+      excludable_items->emplace_back(std::move(copy_or_create_work_item),
+                                     std::move(after_copy_or_create_work_item));
+    } else {
+      ROCKS_LOG_INFO(options_.info_log, "Copying %s to %s", fname.c_str(),
+                     copy_dest_path->c_str());
+      files_to_copy_or_create_.write(std::move(copy_or_create_work_item));
+      backup_items_to_finish.push_back(
+          std::move(after_copy_or_create_work_item));
+    }
```

Comment:
Adding work items and processing them feels a bit more scattered. What if you queue the work items no matter excludable or not (still keeping track of excludable filenames and still invoking the callback in CreateNewBackup*()), then make the worker threads wait for the exclusion list to be ready and let them use it to skip file copies, and then call AddExcludedFile for excluded files in the finishing loop?

Reply:
Right now the file copying can begin while the checkpoint is being taken. I was worried about regressing that by forcing all file copying to wait for the checkpoint to finish, even though it's probably a minor difference.